streams/src/test/java/org/apache/kafka/streams/integration/KStreamRepartitionIntegrationTest.java

```diff
@@ -0,0 +1,548 @@
+                new KeyValue<>(2, "D")
+            )
+        );
+
+        validateReceivedMessages(
+            new StringDeserializer(),
+            new LongDeserializer(),
+            Arrays.asList(
+                new KeyValue<>("1", 2L),
+                new KeyValue<>("2", 2L)
+            )
+        );
+
+        final String repartitionTopicName = toRepartitionTopicName(repartitionName);
+
+        assertTrue(topicExists(repartitionTopicName));
+        assertEquals(2, getNumberOfPartitionsForTopic(repartitionTopicName));
+    }
+
+    private int getNumberOfPartitionsForTopic(final String topic) throws ExecutionException, InterruptedException {
+        try (AdminClient adminClient = createAdminClient()) {
+            final TopicDescription topicDescription = adminClient.describeTopics(Collections.singleton(topic))
+                .values()
+                .get(topic)
+                .get();
+
+            return topicDescription.partitions().size();
+        }
+    }
+
+    private boolean topicExists(final String topic) throws InterruptedException, ExecutionException {
+        try (AdminClient adminClient = createAdminClient()) {
+            final Set<String> topics = adminClient.listTopics()
+                .names()
+                .get();
+
+            return topics.contains(topic);
+        }
+    }
+
+    private String toRepartitionTopicName(final String input) {
+        return applicationId + "-" + input + "-repartition";
+    }
+
+    private AdminClient createAdminClient() {
+        final Properties properties = new Properties();
+        properties.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+
+        return AdminClient.create(properties);
+    }
+
+    private int countOccurrencesInTopology(final String topologyString,
+                                           final String searchPattern) {
+        final Matcher matcher = Pattern.compile(searchPattern).matcher(topologyString);
+        final List<String> repartitionTopicsFound = new ArrayList<>();
+
+        while (matcher.find()) {
+            repartitionTopicsFound.add(matcher.group());
+        }
+
+        return repartitionTopicsFound.size();
+    }
+
+    private void sendEvents(final long timestamp,
+                            final List<KeyValue<Integer, String>> events) throws ExecutionException, InterruptedException {
+        IntegrationTestUtils.produceKeyValuesSynchronouslyWithTimestamp(
+            inputTopic,
+            events,
+            TestUtils.producerConfig(
+                CLUSTER.bootstrapServers(),
+                IntegerSerializer.class,
+                StringSerializer.class,
+                new Properties()
+            ),
+            timestamp
+        );
+    }
+
+    private KafkaStreams startStreams(final StreamsBuilder builder) {
+        final KafkaStreams kafkaStreams = new KafkaStreams(builder.build(streamsConfiguration), streamsConfiguration);
+        kafkaStreams.start();
```

Comment:
Maybe consider:

```suggestion
        final CountDownLatch latch = new CountDownLatch(1);
        kafkaStreams.setStateListener((newState, oldState) -> {
            if (oldState == KafkaStreams.State.REBALANCING && newState == KafkaStreams.State.RUNNING) {
                latch.countDown();
            }
        });
        kafkaStreams.start();
        try {
            latch.await(IntegrationTestUtils.DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS);
        } catch (final InterruptedException e) {
            throw new RuntimeException(e);
        }
```

Then, this method won't return until Streams is actually started, which we've seen can increase test stability.

Reply:
Done